encoder/libav_encoder.cpp

```diff
@@ -0,0 +1,548 @@
+		throw std::runtime_error("libav: cannot find audio encoder " + options->audio_codec);
+
+	codec_ctx_[AudioOut] = avcodec_alloc_context3(codec);
+	if (!codec_ctx_[AudioOut])
+		throw std::runtime_error("libav: cannot allocate audio in context");
+
+	assert(stream_[AudioIn]);
+	codec_ctx_[AudioOut]->channels = stream_[AudioIn]->codecpar->channels;
+	codec_ctx_[AudioOut]->channel_layout = av_get_default_channel_layout(stream_[AudioIn]->codecpar->channels);
+	codec_ctx_[AudioOut]->sample_rate = stream_[AudioIn]->codecpar->sample_rate;
+	codec_ctx_[AudioOut]->sample_fmt = codec->sample_fmts[0];
+	codec_ctx_[AudioOut]->bit_rate = options->audio_bitrate;
+	// usec timebase
+	codec_ctx_[AudioOut]->time_base = { 1, 1000 * 1000 };
+
+	assert(out_fmt_ctx_);
+	if (out_fmt_ctx_->oformat->flags & AVFMT_GLOBALHEADER)
+		codec_ctx_[AudioOut]->flags |= AV_CODEC_FLAG_GLOBAL_HEADER;
+
+	int ret = avcodec_open2(codec_ctx_[AudioOut], codec, nullptr);
+	if (ret < 0)
+		throw std::runtime_error("libav: unable to open audio codec: " + std::to_string(ret));
+
+	stream_[AudioOut] = avformat_new_stream(out_fmt_ctx_, codec);
+	if (!stream_[AudioOut])
+		throw std::runtime_error("libav: cannot allocate stream for audio output context");
+
+	stream_[AudioOut]->time_base = codec_ctx_[AudioOut]->time_base;
+	avcodec_parameters_from_context(stream_[AudioOut]->codecpar, codec_ctx_[AudioOut]);
+}
+
+LibAvEncoder::LibAvEncoder(VideoOptions const *options, StreamInfo const &info)
+	: Encoder(options), init_output_(false), abort_video_(false), abort_audio_(false),
+	  video_start_ts_(0), audio_start_ts_(0), in_fmt_ctx_(nullptr), out_fmt_ctx_(nullptr)
+{
+	avdevice_register_all();
+
+	if (options->verbose)
+		av_log_set_level(AV_LOG_VERBOSE);
+
+	initVideoCodec(options, info);
+	if (options->libav_audio)
+	{
+		initAudioInCodec(options, info);
+		initAudioOutCodec(options, info);
+		av_dump_format(in_fmt_ctx_, 0, options_->audio_device.c_str(), 0);
+	}
+
+	av_dump_format(out_fmt_ctx_, 0, options_->output.c_str(), 1);
+
+	if (options->verbose)
+		std::cerr << "libav: codec init completed" << std::endl;
+
+	video_thread_ = std::thread(&LibAvEncoder::videoThread, this);
+
+	if (options->libav_audio)
+		audio_thread_ = std::thread(&LibAvEncoder::audioThread, this);
+}
+
+LibAvEncoder::~LibAvEncoder()
+{
+	if (options_->libav_audio)
+	{
+		abort_audio_ = true;
+		audio_thread_.join();
+	}
+
+	abort_video_ = true;
+	video_thread_.join();
+
+	avformat_free_context(out_fmt_ctx_);
+	avcodec_free_context(&codec_ctx_[Video]);
+
+	if (options_->libav_audio)
+	{
+		avformat_free_context(in_fmt_ctx_);
+		avcodec_free_context(&codec_ctx_[AudioIn]);
+		avcodec_free_context(&codec_ctx_[AudioOut]);
+	}
+
+	if (options_->verbose)
+		std::cerr << "libav closed" << std::endl;
+}
+
+void LibAvEncoder::EncodeBuffer(int fd, size_t size, void *mem, StreamInfo const &info, int64_t timestamp_us)
+{
+	AVFrame *frame = av_frame_alloc();
+	if (!frame)
+		throw std::runtime_error("could not allocate AVFrame");
```

Comment:
Might be nice if all the errors had a common preamble (I think pretty much everywhere else I see `libav:`), though I agree this one is pretty obvious.

Reply:
Will do!